### ISSUE TITLE
Direct Implementation of K-Nearest neighbor (KNN) in pytorch

### ISSUE THREAD

**Issue by take2rohit** [pytorch/pytorch#71386]
### 🚀 The feature, motivation and pitch

As we all know how important KNN algorithms are in ML research, it would be great to have a direct implementation of the same in PyTorch. I know that though KNN can be coded every time without using many lines of code, I believe that a direct implementation would be time-saving and efficient.

If the devs want then I can create a PR for this.

**Comment by vadimkantorov** [comment]
One usecase I encountered: a useful primitive would be a generalized matmul variant that's can be used for direct knn-graph computation for small-sized vector sets that fit into memory, i.e. given a NxD float32 input database matrix, MxD input query matrix (can be identical to database), a parameter K, return a MxK int64 index matrix wrt dot product / L2. More efficient index / query mechanisms exist in faiss etc, but even the basic one would be useful.

**Comment by take2rohit** [comment]
Hey @vadimkantorov! 
I agree with your point. I feel even a basic implementation for KNN should be there in the vanilla PyTorch library initally. Should I go ahead and create a PR for the same?

**Comment by vadimkantorov** [comment]
yeah, the problem with these is to have a memory efficient implementation (and without fusing matmul + topk this requires to do batching on query side or database side or both. with fusion batching isn't needed for a larger range of input sizes

**Comment by take2rohit** [comment]
Ok I will try to look into resources where people have memory-efficient approach for KNN.

**Comment by vadimkantorov** [comment]
just fyi, i'm not pytorch maintainer, just a fellow user like you :)

**Comment by mruberry** [comment]
I don't think we'd take a KNN implementation in PyTorch itself at this time, but it seems like something libraries built on PyTorch could offer.

**Comment by jbschlosser** [comment]
> One usecase I encountered: a useful primitive would be a generalized matmul variant that's can be used for direct knn-graph computation for small-sized vector sets that fit into memory, i.e. given a NxD float32 input database matrix, MxD input query matrix (can be identical to database), a parameter K, return a MxK int64 index matrix wrt dot product / L2 (and maybe the distance values). More efficient index / query mechanisms exist in faiss etc, but even the basic one would be useful.

I have to agree that this would be useful. Note that this is distinct from e.g. sklearn's [`KNeighborsClassifier`](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html), as that is a classification algorithm and the KNN under discussion would be more of a query for the top-k closest embeddings. While I agree with @mruberry that we don't want to provide the classification algorithm in PyTorch (as it already exists in other libraries such as sklearn), I do think the top-k query version is commonly needed and would compose well with what we provide in PyTorch.

**Comment by bsun0802** [comment]
Exact same use case encountered here, any follow-up on this KNN operator request?

**Comment by jbschlosser** [comment]
Hey @bsun0802, unfortunately there is no one within the PyTorch team working on this at the moment.

**Comment by vadimkantorov** [comment]
Might be a minimal codebase for integrating PQ-based nearest-neighbor retrieval: https://github.com/thomasahle/fast_pq/

**Comment by adv010** [comment]
Hi, Any updates on this?

**Comment by albanD** [comment]
@vadimkantorov do you have examples of where this is used for RAG?

**Comment by vadimkantorov** [comment]
Yeah, classical RAG uses text-based retrieval. I guess, k-NN could be useful for semantic retrieval based on sentence embeddings or similar...

**Comment by vadimkantorov** [comment]
@huguesva maybe some of simpler https://github.com/torchdr/torchdr niceties (like basic k-nn or maybe some benchmarks of building sparse affinities with torch.compile vs keops) can be upstreamed :)

**Comment by JorgeLuizFranco** [comment]
Totally agree. Do we have any hand made implementation that is fast? In the case of nearest k embeddings for example? That is, inside a neural network, one of the steps is to get the KNN and then pass again to other MLP, so we need to do the backpropagation in both MLPs, but not in the KNN
